### ISSUE TITLE
[Bug] Moving item in tree entity removes parent of moved item

### ISSUE THREAD

**Issue by Kampfmoehre** [typeorm/typeorm#5483]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[x] `0.2.22` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

When saving a tree entity where an item was moved to another parent, Typeorm generates 2 queries.  The first one updates the entity to the correct parent, the second one updates the parent again but this time to null.

This kind of applies to deleted items where the parent would be nulled instead of the item beeing deleted.

 I have created an example unit test that fails, see below. I'm not exactly sure, but I guess the save logic traverses through the tree and sees, that the old parent had previously the moved item as child but now doesn't - so it updates the parent of the item to null.
```ts 
import { Entity } from "../../../../src/decorator/entity/Entity";
import { PrimaryGeneratedColumn } from "../../../../src/decorator/columns/PrimaryGeneratedColumn";
import { Tree } from "../../../../src/decorator/tree/Tree";
import { TreeChildren } from "../../../../src/decorator/tree/TreeChildren";
import { TreeParent } from "../../../../src/decorator/tree/TreeParent";

@Entity()
@Tree("nested-set")
export class Node {
    @PrimaryGeneratedColumn()
    public id: number;

    @TreeParent()
    public parent?: Node;

    @TreeChildren({ cascade: true })
    public children: Node[];
}
```
```ts 
import "reflect-metadata";
import {
    createTestingConnections,
    closeTestingConnections,
    reloadTestingDatabases
} from "../../utils/test-utils";
import { Connection } from "../../../src/connection/Connection";
import { expect } from "chai";
import { Node } from "./entity/Node";

describe("github issues > #5483 [Bug] Moving item in tree entity removes parent of moved item", () => {
    let connections: Connection[];
    before(
        async () =>
            (connections = await createTestingConnections({
                entities: [__dirname + "/entity/*{.js,.ts}"],
                schemaCreate: true,
                dropSchema: true
            }))
    );
    beforeEach(() => reloadTestingDatabases(connections));
    after(() => closeTestingConnections(connections));

    it("should update parent correctly when moving", () =>
        Promise.all(
            connections.map(async connection => {
                let root = new Node();
                const child1 = new Node();
                let child12 = new Node();
                const child2 = new Node();
                child1.children = [child12];
                root.children = [child1, child2];

                const repo = connection.getTreeRepository(Node);
                root = await repo.save(root);

                root = await repo.findDescendantsTree(root);
                // move the subchild from child1 to child2
                const movedChilds = root.children[0].children.splice(0, 1);
                root.children[1].children = movedChilds;
                root = await repo.save(root);

                const roots = await repo.findRoots();
                expect(roots).to.have.length(1); // errors with AssertionError: expected [ { id: 1 }, { id: 3 } ] to have a length of 1 but got 2

                root = await repo.findDescendantsTree(root);
                expect(root).to.deep.equal({
                    children: [
                        {
                            children: [],
                            id: child1.id
                        },
                        {
                            children: [
                                {
                                    children: [],
                                    id: child12.id
                                }
                            ],
                            id: child2.id
                        }
                    ],
                    id: root.id
                } as Node);

/** errors with
      AssertionError: expected { Object (children, id) } to deeply equal { Object (children, id) }
      + expected - actual

             "children": []
             "id": 2
           }
           {
      -      "children": []
      +      "children": [
      +        {
      +          "children": []
      +          "id": 3
      +        }
      +      ]
             "id": 4
           }
         ]
         "id": 1
*/
            })
        ));
});
```

I have observed the queries that were generated and saw something like this:
```sql
UPDATE "node" SET "parentId" = $2 WHERE "id" IN ($1) -- PARAMETERS: [3, 4]
UPDATE "node" SET "parentId" = $2 WHERE "id" = $1 -- PARAMETERS: [3, null]
```

I know that moving and deleting is not fully implemented, but apart from the problem save is working pretty well. I have work-around the delete problem with the diff library from FitBit like this:
```ts
import { diff } from "deep-diff";
// ...
const newNodes: Node[] = [];
const deletedNodes: Node[] = [];
const differences = diff(oldRoot, newRoot);

for (const differ of differences) {
  switch (differ.kind) {
    case "A":
      // push new nodes that have an ID, as they are moved nodes
      if (differ.item.kind === "N" && differ.item.rhs.id) {
        newNodes.push(differ.item.rhs);
       } else if (differ.item.kind === "D") {
         deletedNodes.push(differ.item.lhs);
       }
       break;
  }
}

const deleted: Node[] = [];
for (const deleteNode of deletedNodes) {
  // find deleted nodes that have no new node counter, these are really deleted
  if (!newNodes.map((nn) => nn.id).includes(deleteNode.id)) {
    deleted.push(deleteNode);
  }
}

// ...
await this.manager.transaction(async (man: EntityManager) => {
  if (deleted.length > 0) {
    for (const toDelete of deleted) {
      await man.remove(toDelete);
    }
  }

  await man.save(Node, newRoot!);
});
```

But this only solves the deleted items. I also tried to delete the moved items first, but then they are inserted with a new ID instead of the old ones. As far as I know one could create entities and set an ID even if they are auto generated.
```sql
INSERT INTO "node"("nsleft", "nsright", "parentId") VALUES ($1, $2, $3) RETURNING "id", "nsleft", "nsright" -- PARAMETERS: [15,16,6]
```

Does anyone know how to get this done or do I have to implement my own logic for removing/moving.

**Comment by aatecey** [comment]
Hit this myself aswell. Move isn't implemented in typeorm yet and remove doesn't update the nest-set correctly so you have to implement them yourself.

**Comment by AlexMesser** [comment]
This issue was probably fixed in #7156. Please open a new issue if there is still a problem.
